transports/quic/src/stream_map.rs

```diff
@@ -0,0 +1,522 @@
+                self.wake_driver();
+                Either::Right(receiver)
+            }
+        }
+    }
+
+    pub(crate) fn close_reason(&self) -> Result<(), Error> {
+        match self
+            .close_reason
+            .as_ref()
+            .map(|e| Error::ConnectionError(e.clone()))
+        {
+            Some(e) => Err(e),
+            None => Ok(()),
+        }
+    }
+
+    pub(crate) fn destroy_stream(&mut self, stream: StreamId) {
+        trace!("Removing substream {:?} from map", *stream);
+        self.connection.destroy_stream(*stream);
+        self.map.remove(&stream.0).expect(
+            "Internal state corrupted. \
+             You probably used a Substream with the wrong StreamMuxer",
+        );
+    }
+
+    pub(crate) fn close(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Error>> {
+        trace!(
+            "close() called with {} outbound streams for side {:?}",
+            self.connection.send_streams(),
+            self.connection.side()
+        );
+        self.wake_closer();
+        self.close_waker = Some(cx.waker().clone());
+        if self.close_reason.is_some() {
+            return Poll::Ready(Ok(()));
+        } else if self.connection.send_streams() == 0 {
+            self.shutdown(0);
+            self.close_reason = Some(quinn_proto::ConnectionError::LocallyClosed);
+            return Poll::Ready(Ok(()));
+        }
+        for (&stream, value) in &mut self.map {
+            value.wake_all();
+            let _ = self.connection.finish(stream);
+        }
+        Poll::Pending
+    }
+
+    pub(crate) fn handle_event(
+        &mut self,
+        event: quinn_proto::ConnectionEvent,
+    ) -> Option<quinn_proto::EndpointEvent> {
+        self.connection.handle_event(event)
+    }
+
+    pub(crate) fn shutdown_stream(
+        &mut self,
+        cx: &mut Context<'_>,
+        substream: &StreamId,
+    ) -> Result<oneshot::Receiver<()>, quinn_proto::FinishError> {
+        self.wake_driver();
+        self.finish(substream)?;
+        let (sender, mut receiver) = oneshot::channel();
+        let _ = receiver.poll_unpin(cx);
+        self.get(substream).set_finisher(sender);
+        Ok(receiver)
+    }
+}
+
+/// The connection driver
+#[derive(Debug)]
+struct ConnectionDriver {
+    inner: Arc<Mutex<Streams>>,
+    /// The packet awaiting transmission, if any.
+    outgoing_packet: Option<quinn_proto::Transmit>,
+    /// The timer being used by this connection
+    timer: Option<futures_timer::Delay>,
+    /// The last timeout returned by `quinn_proto::poll_timeout`.
+    last_timeout: Option<Instant>,
+    /// Transmit socket
+    socket: Arc<socket::Socket>,
+}
+
+impl Upgrade {
+    pub(crate) fn spawn(
+        connection: Arc<Mutex<Streams>>,
+        _side: quinn_proto::Side,
+        socket: Arc<crate::socket::Socket>,
+    ) -> Upgrade {
+        let inner = connection.clone();
+        async_std::task::spawn(ConnectionDriver {
```

Comment:
Spawning a task per connection should not be necessary and unbounded task creation means unbounded resource use. Instead the muxer should make progress via polling. Sadly the current `StreamMuxer` trait does not define a general poll method, but the behaviour can be approximated by making progress via `poll_inbound` which should then also do what `process_app_events` does currently.

Reply:
When writing this code, I was not aware that `poll_inbound` was guaranteed to be called regularly, so I used a background task instead.  Will fix.

Reply:
> [...] behaviour can be approximated by making progress via `poll_inbound` [...]

That should have read "via `poll_inbound` and `close`", cf. [here](https://github.com/libp2p/rust-libp2p/blob/96cc7912a3f885ac9a09ed6eb31ae8854c278eee/muxers/yamux/src/lib.rs#L106) and [here](https://github.com/libp2p/rust-libp2p/blob/96cc7912a3f885ac9a09ed6eb31ae8854c278eee/muxers/yamux/src/lib.rs#L153).